File: src/main/java/brightspark/runicmagic/enums/SpellType.java
package brightspark.runicmagic.enums;

import net.minecraft.client.resources.I18n;
import net.minecraft.util.text.TextFormatting;

import java.util.Locale;

import static brightspark.runicmagic.enums.SpellType.MagicType.*;

public enum SpellType
{
    ELEMENTAl(COMBAT),
    CURSES(COMBAT),
    SUPPORT(COMBAT),

    TELESELF(TELEPORT),
    TELEOTHER(TELEPORT),
    TELEGROUP(TELEPORT),

    ENCHANTMENT(SKILLING),
    ALCHEMY(SKILLING),
    OTHER(SKILLING);

    private final MagicType magicType;
    private final String unlocName;

    SpellType(MagicType magicType)
    {
        this.magicType = magicType;
        unlocName = "spelltype." + name().toLowerCase(Locale.ROOT) + ".name";
    }

    public MagicType getMagicType()
    {
        return magicType;
    }

    public String getTranslation()
    {
        return I18n.format(unlocName);
    }

    public enum MagicType
    {
        COMBAT(TextFormatting.RED),
        TELEPORT(TextFormatting.BLUE),
        SKILLING(TextFormatting.GOLD);

        private final TextFormatting textColour;

        MagicType(TextFormatting textColour)
        {
            this.textColour = textColour;
        }

        public TextFormatting getTextColour()
        {
            return textColour;
        }
    }
}

File: src/main/kotlin/brightspark/runicmagic/model/SpellType.kt
package brightspark.runicmagic.model

import brightspark.runicmagic.model.SpellType.MagicType.*
import net.minecraft.client.resources.I18n
import net.minecraft.util.text.TextFormatting
import java.util.*

enum class SpellType(val magicType: MagicType) {
	ELEMENTAL(COMBAT),
	CURSES(COMBAT),
	SUPPORT(COMBAT),

	TELESELF(TELEPORT),
	TELEOTHER(TELEPORT),
	TELEGROUP(TELEPORT),

	ENCHANTMENT(SKILLING),
	ALCHEMY(SKILLING),
	OTHER(SKILLING);

	val unlocName: String = "spelltype.${name.toLowerCase(Locale.ROOT)}.name"

	fun getTranslation(): String = I18n.format(unlocName)

	fun getMagicTypeColour(): TextFormatting = magicType.textColour

	enum class MagicType(val textColour: TextFormatting) {
		COMBAT(TextFormatting.RED),
		TELEPORT(TextFormatting.BLUE),
		SKILLING(TextFormatting.GOLD)
	}
}
